user-fund-storage/src/fund_module.rs

```diff
@@ -303,11 +323,63 @@ pub trait FundModule {
                     (*fund_item).user_id,
                     transformed,
                     extracted_balance);
-                affected_users.push((*fund_item).user_id)
+                
+            }
+            id = next_id;
+        }
+
+        affected_users.sort();
+        affected_users.dedup();
+        affected_users
+    }
+
+    /// Same as split_convert_max_by_type, but does not change any of the 
+    fn simulate_split_convert_max_by_type<F, I>(&self,
+        mut opt_max_amount: Option<&mut BigUint>,
+        source_type: FundType,
+        direction: SwapDirection,
+        mut filter: F,
+        interrupt: I) -> Vec<usize>
+    where 
+        F: FnMut(usize, FundDescription) -> bool,
+        I: Fn() -> bool
+    {
+        let type_list = self.get_fund_list_by_type(source_type);
+        let mut affected_users: Vec<usize> = Vec::new();
+        let mut id = match direction {
+            SwapDirection::Forwards => type_list.first,
+            SwapDirection::Backwards => type_list.last,
+        };
+
+        while id > 0 && !interrupt() {
+            if let Some(max_amount) = &opt_max_amount {
+                if **max_amount == 0 {
+                    break; // do not process anything after the max_amount is completely drained
+                }
+            }
+
+            let fund_item = self.get_mut_fund_by_id(id);
+            affected_users.push(fund_item.user_id);
+            let next_id = match direction { // save next id now, because fund_item can be destroyed
+                SwapDirection::Forwards => fund_item.type_list_next,
+                SwapDirection::Backwards => fund_item.type_list_prev,
+            };
+
+            if filter(fund_item.user_id, fund_item.fund_desc) {
```

Comment:
you need only this part from 368-376 from the simulate function inside the first one which does the actual split/swap. If you get the action of swap out into a function - and the max_amount computation into one as well. Thus in case of simulate - the second function does not have to be called. 
Doing this there is no need to duplicate code.